langpuppet/gapi.go

```diff
@@ -0,0 +1,290 @@
+		},
+	}
+
+	dataPuppet := &gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(puppet.Name+": "+format, v...)
+		},
+	}
+
+	obj.langGAPI = &lang.GAPI{}
+	if err := obj.langGAPI.Init(dataLang); err != nil {
+		return err
+	}
+	obj.puppetGAPI = &puppet.GAPI{}
+	if err := obj.puppetGAPI.Init(dataPuppet); err != nil {
+		return err
+	}
+
+	obj.closeChan = make(chan struct{})
+	obj.initialized = true
+	return nil
+}
+
+// Graph returns a current Graph.
+func (obj *GAPI) Graph() (*pgraph.Graph, error) {
+	if !obj.initialized {
+		return nil, fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	if obj.currentLangGraph == nil || obj.currentPuppetGraph == nil {
+		return nil, fmt.Errorf("%s: graph was requested before both Lang and Puppet graphs were available", Name)
+	}
+
+	lg := obj.currentLangGraph
+	pg := obj.currentPuppetGraph
+
+	g := &pgraph.Graph{
+		Name: lg.Name + "x" + pg.Name,
+	}
+
+	return g, err
+}
+
+// Next returns nil errors every time there could be a new graph.
+func (obj *GAPI) Next() chan gapi.Next {
+	ch := make(chan gapi.Next)
+	obj.wg.Add(1)
+	go func() {
+		defer obj.wg.Done()
+		defer close(ch) // this will run before the obj.wg.Done()
+		if !obj.initialized {
+			next := gapi.Next{
+				Err:  fmt.Errorf("%s: GAPI is not initialized", Name),
+				Exit: true, // exit, b/c programming error?
+			}
+			ch <- next
+			return
+		}
+		nextLang := obj.langGAPI.Next()
+		nextPuppet := obj.puppetGAPI.Next()
+
+		var currentLangGraph *pgraph.Graph
+		var currentPuppetGraph *pgraph.Graph
+
+		for {
+			var graph_err error
+
+			select {
+			case nextChild <- nextLang:
+				graph_err = nextChild.Err
+				if nextChild.Err == nil {
+					obj.currentLangGraph, graph_err = obj.langGAPI.Graph()
+				}
+			case nextChild <- nextPuppet:
+				graph_err = nextChild.Err
+				if nextChild.Err == nil {
+					obj.currentPuppetGraph, graph_err = obj.puppetGAPI.Graph()
+				}
+			case <-obj.closeChan:
+				return
+			}
+
+			var next gapi.Next
+
+			if graph_err != nil {
+				next = gapi.Next{
+					Err: graph_err,
+				}
+			}
+
+			if obj.currentLangGraph == nil || obh.currentPuppetGraph == nil {
+				continue
+			}
+
+			obj.data.Logf("generating new composite graph...")
+			next := gapi.Next{
+				//Exit: true, // TODO: for permanent shutdown!
+				Err: nil,
+			}
+			select {
+			case ch <- next: // trigger a run (send a msg)
+			// unblock if we exit while waiting to send!
+			case <-obj.closeChan:
+				return
+			}
+		}
+	}()
+	return ch
+}
+
+// Close shuts down the Puppet GAPI.
+func (obj *GAPI) Close() error {
+	if !obj.initialized {
+		return fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	var closerGroup sync.WaitGroup
```

Comment:
Not sure you need goroutines in close. It can just be linear/sync. Not sure there's much of an added benefit, and it makes error passing harder which you're not currently doing. (Not that the errors in Close are very important.)

Reply:
Done.